### PR TITLE
feat: aesthetic laion metric

### PR DESCRIPTION
**Pull request by begumcig**
## Description

This PR introduces AestheticLAION evaluation metric to Pruna, based on  [[LAION-AI aesthetic-predictor](https://github.com/LAION-AI/aesthetic-predictor)](https://github.com/LAION-AI/aesthetic-predictor).

The implementation was contributed by [kirdmiv](https://github.com/kirdmiv) in PR https://github.com/PrunaAI/pruna/pull/384 and this branch brings their work into the main PrunaAI repository. 

Full credit for the original implementation goes to @kirdmiv 💜💜💜 Thanks a lot for your amazing contribution!

## Related Issue

Fixes #(issue number)

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: src/pruna/evaluation/metrics/aesthetic_laion.py [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+    ) -> None:
+        super().__init__(*args, **kwargs)
+        if model_name_or_path not in self.model_name_to_aesthetic_head_name:
+            pruna_logger.error(f"Model {model_name_or_path} does not exist.")
+            raise ValueError(f"Model {model_name_or_path} does not exist.")
+
+        self.device = set_to_best_available_device(device)
+        try:
+            model_info(model_name_or_path)
+        except (EntryNotFoundError, RepositoryNotFoundError):
```
Comment (davidberenstein1957) [inline src/pruna/evaluation/metrics/aesthetic_laion.py]: why are we adding this if it should never happen/ also we can just rely on the from pretrained methods lateron, right?

File: src/pruna/evaluation/metrics/aesthetic_laion.py [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+            A pretrained linear head corresponding to the given CLIP model name.
+        """
+        home = pathlib.Path("~").expanduser()
+        cache_folder = home / ".cache/aesthetic_laion_linear_heads"
+        clip_model = self.model_name_to_aesthetic_head_name[clip_model]
+        path_to_model = cache_folder / ("sa_0_4_" + clip_model + "_linear.pth")
+        if not path_to_model.exists():
+            cache_folder.mkdir(exist_ok=True, parents=True)
+            url_model = (
+                "https://github.com/LAION-AI/aesthetic-predictor/blob/main/sa_0_4_" + clip_model + "_linear.pth?raw=true"
```
Comment (davidberenstein1957) [inline src/pruna/evaluation/metrics/aesthetic_laion.py]: given that this is harcoded can't we just allow for the 3 models mentioned above?

File: src/pruna/evaluation/metrics/aesthetic_laion.py [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+
+            see https://github.com/LAION-AI/aesthetic-predictor/tree/main for available variants
+
+        Returns
+        -------
+        torch.nn.Linear
+            A pretrained linear head corresponding to the given CLIP model name.
+        """
+        home = pathlib.Path("~").expanduser()
+        cache_folder = home / ".cache/aesthetic_laion_linear_heads"
```
Comment (davidberenstein1957) [inline src/pruna/evaluation/metrics/aesthetic_laion.py:169]: why are we setting this manually?

File: tests/evaluation/test_aesthetics_laion.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+        pytest.param("cpu", "openai/clip-vit-large-patch14", marks=pytest.mark.cpu),
+        pytest.param("cpu", "openai/clip-vit-base-patch32", marks=pytest.mark.cpu),
+        pytest.param("cpu", "openai/clip-vit-base-patch16", marks=pytest.mark.cpu),
+        pytest.param("cuda", "openai/clip-vit-large-patch14", marks=pytest.mark.cuda),
+    ],
+)
+def test_aesthetic_laion(device: str, clip_model: str) -> None:
+    """Test the AestheticLAION metric."""
+    data_module = PrunaDataModule.from_string("LAION256")
+    data_module.limit_datasets(10)
```
Comment (davidberenstein1957) [inline tests/evaluation/test_aesthetics_laion.py]: perhaps a smaller size is enough? do we want to test inference with all models? perhaps just checking if they exist and we can download them is enough?

Comment (begumcig) [reply tests/evaluation/test_aesthetics_laion.py]: I think the small models are small enough for it to be OK to run on CPU! Good point on the number of points, I am decreasing it!

File: src/pruna/evaluation/metrics/aesthetic_laion.py (hunk 1 above)
Comment (begumcig) [reply src/pruna/evaluation/metrics/aesthetic_laion.py]: Good point!

File: src/pruna/evaluation/metrics/aesthetic_laion.py (hunk 2 above)
Comment (begumcig) [reply src/pruna/evaluation/metrics/aesthetic_laion.py]: Yes, updated! Lmk what you think :)

File: src/pruna/evaluation/metrics/aesthetic_laion.py (hunk 3 above)
Comment (begumcig) [reply src/pruna/evaluation/metrics/aesthetic_laion.py:169]: I found this quick and easy to check but honestly I don't know much better, do you have a suggestion?

Comment (davidberenstein1957) [reply src/pruna/evaluation/metrics/aesthetic_laion.py:169]: There are methods like ` hf_hub_download` or `snapshot` download which return the path of the snapshot or file that you retrieved, perhaps that helps. Also, if the file is already there it will skip it and just return the path, perhaps that is cleaner?

Comment (begumcig) [reply src/pruna/evaluation/metrics/aesthetic_laion.py:169]: ok that sounds insanely cool, looking into it!

Comment (begumcig) [reply src/pruna/evaluation/metrics/aesthetic_laion.py:169]: Okay, I looked into it, seems to be designed for repos on huggingface, and we are downloading these predictors from github, shall we skip this for now?

Comment (davidberenstein1957) [reply src/pruna/evaluation/metrics/aesthetic_laion.py:169]: ahh, sorry, I missed this. If it is just for GH we can skip my comments.

**Review by davidberenstein1957 — APPROVED** [review]
I did not realise we were downloading from GH. I did not say anything, bt still the `hf_hub_download` etc methods are coooool indeed. 👾
